Windows Backup Privilege Tweaks

Crazycatz00 opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

This PR enables backup semantics for file EA opens and attempts to enable `SeBackupPrivilege` when first beginning a backup.

When enabled, `SeBackupPrivilege` allows read-only opening that bypasses ACLs; this can permit a restricted account with the proper privilege to back up all users' files without changing file-system permissions. Go supports backup semantics for read-only opens since [405275 / v1.20](<https://go-review.googlesource.com/c/go/+/405275>) and restic currently enables the privilege to backup security descriptors, so such backups _mostly_ work already, with two issues:
1. Reading extended attributes fails for files when `SeBackupPrivilege` allows access which would have been denied without the privilege.
   This is because `windows.CreateFile` is used directly without the appropriate flag.
   (Directories already have the flag as they require it to get a handle at all.)
2. `SeBackupPrivilege` is only enabled upon reading the first security descriptor; accesses before this are subject to normal ACLs and may fail. This always affects the scanner, and the actual backup will error & skip initial items until one has permissive enough ACLs.

An example of both issues:
```console
> rem Running in cmd.exe as Administrator
> set RESTIC_PASSWORD=0000
> restic init -r ".\repo"
> mkdir ".\data"

> rem Create directory "b" accessible only to the SYSTEM account
> mkdir ".\data\b" && echo Test > ".\data\b\b.txt" && icacls ".\data\b" /inheritance:r /grant SYSTEM:F

> restic backup -r ".\repo" -nvv ".\data\b"
open repository
repository 5403b32f opened (version 2, compression level auto)
no parent snapshot found, will read all files
load index files
[0:00]          0 index files loaded
start scan on [.\data\b]
start backup on [.\data\b]
scan: openfile for readdirnames failed: open \\?\D:\Test\data\b: Access is denied.
scan finished in 0.003s: 0 files, 0 B
error: incomplete metadata for data\b\b.txt: get EA failed while opening file handle for path data\b\b.txt, with: Access is denied.
new       /data/b/b.txt, saved in 0.003s (8 B added)
new       /data/b/, saved in 0.004s (0 B added, 0 B stored, 412 B metadata)
new       /data/, saved in 0.007s (0 B added, 0 B stored, 429 B metadata)

Files:           1 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Data Blobs:      1 new
Tree Blobs:      3 new
Would add to the repository: 1.778 KiB (1.380 KiB stored)

processed 1 files, 8 B in 0:00
Warning: at least one source file could not be read

> cd ".\data" && restic backup -r "..\repo" -nvv "b"
open repository
repository 5403b32f opened (version 2, compression level auto)
no parent snapshot found, will read all files
load index files
[0:00]          0 index files loaded
start scan on [b]
start backup on [b]
scan: openfile for readdirnames failed: open \\?\D:\Test\data\b: Access is denied.
scan finished in 0.001s: 0 files, 0 B
error: openfile for readdirnames failed: open \\?\D:\Test\data\b: Access is denied.
Fatal: unable to save snapshot: snapshot is empty
```


Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

No.


Checklist
---------

- [x] I have added tests.
- ~[ ] I have added documentation for relevant changes (in the manual).~
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

Commit:
fs: Use backup privileges when reading extended attributes for files too.

jtru commented:
This is a great feature that I would very much like to have asap, but I *think* it would be better to have it as an explicit opt-in. Otherwise, I expect all kinds of "endpoint protection" and anti-virus snake oil  to trip over `restic` executables in the near future, which would be a real hassle. Better let users explicitly enable the backup privilege in case they know they may actually attain it.

MichaelEischer commented on cmd/restic/cmd_backup.go:
```diff
@@ -482,6 +482,7 @@ func runBackup(ctx context.Context, opts BackupOptions, gopts GlobalOptions, ter
 	var err error
 
 	if runtime.GOOS == "windows" {
+		fs.EnableBackupPrivilege()
```
This doesn't compile on non-windows systems. I'd prefer to let `internal/fs/fs_local.go` handle the privilege request automatically. It should request the privileges when the fs instance is created.

MichaelEischer commented on internal/fs/sd_windows.go:
```diff
@@ -42,7 +42,7 @@ var lowRestoreSecurityFlags windows.SECURITY_INFORMATION = windows.DACL_SECURITY
 // If there are no admin permissions, only the current user's owner, group and DACL will be got.
 func getSecurityDescriptor(filePath string) (securityDescriptor *[]byte, err error) {
-	onceBackup.Do(enableBackupPrivilege)
+	onceBackup.Do(EnableBackupPrivilege)
```
If `EnableBackupPrivilege` is already called by the backup command, then this call here is redundant.

MichaelEischer reviewed (changes requested):
I'd like to keep the privilege management hidden in the fs package, see below for a suggestion how to achieve that.

Commit:
fs: Attempt to enable file system privileges on initialization.

Add tests to verify privileges' effects.

Crazycatz00 commented:
This should be a bit more proper. Contained to the `internal/fs` module, running just once on `init()`, and made sure it builds for at least Linux via `GOOS`. Also rebased to latest and added two tests to verify the behavior.

MichaelEischer commented on internal/fs/priv_windows_test.go:
```diff
@@ -0,0 +1,61 @@
+	test.OK(t, testfile.Close())
+
+	// Set restricted permissions.
+	sd, err := windows.SecurityDescriptorFromString("D:PAI(D;;FRFWFX;;;WD)(A;;SD;;;WD)")
```
Please add a short comment what the security descriptor does. I just see a wild collection of characters ^^

MichaelEischer reviewed (changes requested):
LGTM, I just have two small nits

Commit:
fs: Clarified documentation

MichaelEischer reviewed (approved):
LGTM
